Revise modular pipelines docs (#3948)
kedro-org/kedro: docs/source/nodes_and_pipelines/pipeline_introduction.md

# Pipelines

> A pipeline is a set of [Nodes](./nodes.md) that will be automatically resolved after pipeline creation to determine the node execution order based on outputs and inputs of each node.

Reviewer: Not sure if this sentence reads very clear to me. I don't really mind the introduction from before but simplified just a little bit?

Technically {py:class}`~kedro.pipeline.Pipeline` is a python class that could be created using {py:class}`~kedro.pipeline.modular_pipeline.pipeline` method, based on Nodes or other Pipelines(in that case all nodes from that pipeline will be used).

Most important about kedro pipelines:
- [How to create a simple pipeline](#how-to-create-a-simple-pipeline)
- [Create a new blank pipeline with kedro pipeline create command](#create-a-new-blank-pipeline-with-the-kedro-pipeline-create-command)
- [Pipeline creation structure](#pipeline-creation-structure)
- [Key Pipeline methods and operations](#key-pipeline-methods-and-operations)
- [Bad pipelines - rules](#bad-pipelines)

> - [Custom new pipeline templates](#custom-new-pipeline-templates)

Reviewer: I think these headings could be all "how to" headlines like we have in configuration docs. https://docs.kedro.org/en/stable/configuration/advanced_configuration.html#advanced-configuration-for-kedro-projects

# Pipeline objects

We previously introduced [Nodes](./nodes.md) as building blocks that represent tasks, and can be combined in a pipeline to build your workflow. A pipeline organises the dependencies and execution order of your collection of nodes, and connects inputs and outputs while keeping your code modular. The pipeline resolves dependencies to determine the node execution order, and does *not* necessarily run the nodes in the order in which they are passed in.

> To benefit from Kedro's automatic dependency resolution, you can chain your nodes into a {py:class}`~kedro.pipeline.Pipeline`, which is a list of nodes that use a shared set of variables. That class can be created using the {py:class}`~kedro.pipeline.modular_pipeline.pipeline` method, based on nodes or other pipelines (in which case all nodes from that pipeline will be used).

Reviewer 2: ```suggestion
To benefit from Kedro's automatic dependency resolution is that you can chain your nodes into a {py:class}`~kedro.pipeline.Pipeline`, which is a list of nodes that use a shared set of variables. That class can be created using the {py:class}`~kedro.pipeline.modular_pipeline.pipeline` method, based on nodes or other pipelines (in which case all nodes from that pipeline will be used).
```

We previously introduced [Nodes](./nodes.md) as building blocks that represent tasks, and can be combined in a pipeline to build your workflow. A pipeline organises the dependencies and execution order of your collection of nodes, and connects inputs and outputs while keeping your code modular. The pipeline resolves dependencies to determine the node execution order, and does *not* necessarily run the nodes in the order in which they are passed in.

To benefit from Kedro's automatic dependency resolution, you can chain your nodes into a {py:class}`~kedro.pipeline.Pipeline`, which is a list of nodes that use a shared set of variables. That class could be created using {py:class}`~kedro.pipeline.modular_pipeline.pipeline` method, based on nodes or other pipelines (in which case all nodes from that pipeline will be used).

> - [How to create a simple pipeline](#how-to-create-a-simple-pipeline)

Reviewer 3: I actually think this is a good suggestion to keep the original wording.

    ]
)
```

> 

Reviewer: A little explanation here about how Kedro can figure out the order of execution of these nodes based on the inputs and outputs of the nodes in a pipeline.

## Create a new blank pipeline with the `kedro pipeline create` command

> When managing your Kedro project, we recommend organising your pipelines separately to achieve modularity. This approach allows for easy copying and reuse of pipelines within and between projects. Simply put: one pipeline, one folder. To help with that, you can use the following command (the pipeline name must adhere to [Python convention](https://realpython.com/python-pep8/#naming-conventions)):

Reviewer: This also reads a little bit unclear to me. I think this part could explain the idea behind dividing tasks into pipelines. For eg, a project usually contains a lot of tasks and we recommend grouping tasks that frequently are executed together into individual pipelines. These pipelines should ideally be organised in separate folders to encourage reuse and modularity.

```bash
kedro pipeline create <pipeline_name>
```

After running this command, a new pipeline with boilerplate folders and files will be created in your project. For your convenience, Kedro gives you a pipeline-specific `nodes.py`, `pipeline.py`, parameters file and appropriate `tests` structure. It also adds the appropriate `__init__.py` files. You can see the generated folder structure below:

```text
├── conf
│   └── base
│       └── parameters_{{pipeline_name}}.yml  <-- Pipeline-specific parameters
└── src
    ├── my_project
    │   ├── __init__.py
    │   └── pipelines
    │       ├── __init__.py
    │       └── {{pipeline_name}}      <-- This folder defines the pipeline
    │           ├── __init__.py        <-- So that Python treats this pipeline as a module
    │           ├── nodes.py           <-- To declare your nodes
    │           └── pipeline.py        <-- To structure the pipeline itself
    └── tests
        ├── __init__.py
        └── pipelines
            ├── __init__.py
            └── {{pipeline_name}}      <-- Pipeline-specific tests
                ├── __init__.py
                └── test_pipeline.py

```

> If you want to do the reverse and remove that pipeline, you can use `kedro pipeline delete <pipeline_name>` to do so or have a look to all [project-specific CLI commands list](../development/commands_reference.md#kedro-commands).

Reviewer: ```suggestion
If you want to delete an existing pipeline, you can use `kedro pipeline delete <pipeline_name>` to do so. 
```

Before filling `pipeline.py` with nodes, we recommend storing all node functions in `nodes.py`. From our previous example, we should put:

```python
# src/my_project/pipelines/{{pipeline_name}}/nodes.py
def mean(xs, n):
    return sum(xs) / n

def mean_sos(xs, n):
    return sum(x**2 for x in xs) / n

def variance(m, m2):
    return m2 - m * m
```

```python
# src/my_project/pipelines/{{pipeline_name}}/pipelines.py
from kedro.pipeline import Pipeline, pipeline, node

from .nodes import mean, mean_sos, variance
# Import node functions from nodes.py located in the same folder

def create_pipeline(**kwargs) -> Pipeline:
    return pipeline(
    [
        node(len, "xs", "n"),
        node(mean, ["xs", "n"], "m", name="mean_node", tags = "tag1"),
        node(mean_sos, ["xs", "n"], "m2", name="mean_sos", tags = ["tag1", "tag2"]),
        node(variance, ["m", "m2"], "v", name="variance_node"),
    ], # A list of nodes and pipelines combined into a new pipeline
    tags = "tag3", # Optional, each pipeline node will be tagged
    namespace = "", # Optional
    inputs = {}, # Optional
    outputs = {}, # Optional

>     parameters = {}, # Optional

Reviewer: Might be missing a closing paranthesis?

    ]
)
```

Kedro determines the order of execution of these nodes based on the inputs and outputs specified for each node. In this example:
1. The first node computes the length of `xs` and outputs it as `n`.
2. The second node calculates the mean using `xs` and `n`, and outputs it as `m`.
3. The third node computes the mean sum of squares (mean_sos) using `xs` and `n`, and outputs it as `m2`.
4. The fourth node calculates the variance using the mean (`m`) and mean sum of squares (`m2`), and outputs it as `v`.

> Kedro's dependency resolution ensures that each node runs only after its required inputs are available from the outputs of previous nodes. This way, the nodes are executed in the correct order automatically, based on the defined dependencies.

Reviewer 3: Maybe it helps to print `pipline.nodes`  which is toposorted already?

Author: It's the section below titled "How to receive information about the nodes in a pipeline" that shows that output, so I don't think it's a good idea to duplicate it on the same page.

    ]
)
```

Kedro determines the order of execution of these nodes based on the inputs and outputs specified for each node. In this example:
1. The first node computes the length of `xs` and outputs it as `n`.
2. The second node calculates the mean using `xs` and `n`, and outputs it as `m`.
3. The third node computes the mean sum of squares (mean_sos) using `xs` and `n`, and outputs it as `m2`.
4. The fourth node calculates the variance using the mean (`m`) and mean sum of squares (`m2`), and outputs it as `v`.

> Kedro's dependency resolution ensures that each node runs only after its required inputs are available from the outputs of previous nodes. This way, the nodes are executed in the correct order automatically, based on the defined dependencies.

Reviewer 4: ```suggestion
Kedro's dependency resolution algorithm ensures that each node runs only after its required inputs are available from the outputs of previous nodes. This way, the nodes are executed in the correct order automatically, based on the defined dependencies.
```

?

## How to use `describe` to discover what nodes are part of the pipeline

> ```python

Reviewer 4: Maybe one line explaining what this is, even if it's short

Outputs: v
##################################
```
</details>

> ### How to tag a pipeline

Reviewer: Not sure why this section is removed from the page?

This is because `.` has a special meaning internally and indicates a namespace pipeline. In the example, the outputs segment should be disconnected as the name implies there is an "output1" namespace pipeline. The input is not namespaced, but the output is via its dot notation. This leads to Kedro processing each separately. For this example, a better approach would've been writing both as `input1_kedro` and `output1_kedro`.

We recommend use of characters like `_` instead of `.` as name separators.

## How to create a new blank pipeline using the `kedro pipeline create` command

> When managing your Kedro project, we recommend grouping related tasks into individual pipelines to achieve modularity. A project typically contains many tasks, and organising frequently executed tasks together into separate pipelines helps maintain order and efficiency. Each pipeline should ideally be organised in its own folder, promoting easy copying and reuse within and between projects. Simply put: one pipeline, one folder. To assist with this, you can use the following command (the pipeline name must adhere to [Python naming conventions](https://realpython.com/python-pep8/#naming-conventions)):

Reviewer 4: From here I'd direct readers to the next page, explaining Modular Pipelines (see review comment)

```suggestion
When managing your Kedro project, we recommend grouping related tasks into individual pipelines to achieve modularity. A project typically contains many tasks, and organising frequently executed tasks together into separate pipelines helps maintain order and efficiency. Each pipeline should ideally be organised in its own folder, promoting easy copying and reuse within and between projects. Simply put: one pipeline, one folder. To assist with this, Kedro introduces the concept of [Modular Pipelines](./modular_pipelines), which are described in the next section.
```

This is because `.` has a special meaning internally and indicates a namespace pipeline. In the example, the outputs segment should be disconnected as the name implies there is an "output1" namespace pipeline. The input is not namespaced, but the output is via its dot notation. This leads to Kedro processing each separately. For this example, a better approach would've been writing both as `input1_kedro` and `output1_kedro`.

We recommend use of characters like `_` instead of `.` as name separators.

## How to store pipeline code in a kedro project

> When managing your Kedro project, we recommend grouping related tasks into individual pipelines to achieve modularity. A project typically contains many tasks, and organising frequently executed tasks together into separate pipelines helps maintain order and efficiency. Each pipeline should ideally be organised in its own folder, promoting easy copying and reuse within and between projects. Simply put: one pipeline, one folder. To assist with this, Kedro introduces the concept of [Modular Pipelines](modular_pipelines.md), which are described in the next section.

Reviewer 5: Probably not `between projects` because of micro packaging deprecation?
